### PR TITLE
Implement Fusion definition equality check

### PR DESCRIPTION
**Pull request by rdspring1**
This PR creates a check to determine if two separate `Fusion` objects have the same definition unlike the existing equality IR checks, which require the object be exactly the same. This second part of creating an `LRU Cache` for direct bindings.

## `sameDefinition` Details
`Val::sameDefinition` checks the Val's definition. `Expr::sameDefinition` checks the Expr's input arguments.
`Fusion::sameDefinition` starts from each output and traverses the entire Fusion DAG. Unlike the legacy `TensorRecord` check, `TensorView::sameDefinition` directly compares the logical and allocation domain, which handles non-trival operations like `preprocessGroupedMatmulInputSf`.

## Differences between `Val::sameAs` and `Val::sameDefinition`
* `Val::sameVal` contains common code for `Val::sameAs` and `Val::sameDefinition`
* `sameAs` determines if a Statement has the exact same result as the other value.
* For `sameAs`, two values without definition nor value are not equal. This is used in concretization.
In addition, if either Val has a non-deterministic definition, then the Vals are not the same.
* `sameDefinition` determines if a two values will create the same fusion definition.
* Symbolic values and NaNs are considered the same.
* The value's integer `name` is checked to ensure argument order.

## Differences between `Expr::sameAs` and `Expr::sameDefinition`
* `Expr::sameDefinition` skips the `isDeterministic` checks.

## PR Stack
- #5453 
- #5454 **<<< This PR.**
- #4893

### CODE REVIEW

**Comment by rdspring1** [comment]
!test

**Comment by jjsjann123** [comment]
I've been a slacker. Let me jump on the review.

File: csrc/ir/base_nodes.cpp [hunk 1]
```diff
@@ -116,32 +116,60 @@ bool Val::removeUse(Expr* expr) {
...
+  }
+  if (value_.hasValue() != other_val->value_.hasValue()) {
+    return false;
+  }
+  if (definition_ == nullptr) {
+    // Check this condition in Val::sameAs and Val::sameDefinition
+    return true;
+  }
+
+  NVF_ERROR(!value_.hasValue());
```
Comment (jacobhinkle) [inline csrc/ir/base_nodes.cpp:143]: Why must this be true?

File: csrc/ir/base_nodes.cpp [hunk 2]
```diff
@@ -116,32 +116,60 @@ bool Val::removeUse(Expr* expr) {
...
+  }
+  if (dtype_ != other_val->dtype_) {
+    return false;
+  }
+  if (value_.hasValue() != other_val->value_.hasValue()) {
+    return false;
+  }
+  if (definition_ == nullptr) {
+    // Check this condition in Val::sameAs and Val::sameDefinition
+    return true;
```
Comment (jjsjann123) [inline csrc/ir/base_nodes.cpp:140]: nitpick: returning `true` for sameAs when this->definition_ == nullptr felt very strange and wrong. I think we should refactor this utility to `notSameVal` instead.

File: csrc/ir/base_nodes.cpp [hunk 3]
```diff
@@ -152,23 +180,53 @@ bool Val::sameAs(const Statement* other) const {
...
-    }
     return true;
   }
   return false;
 }
 
+namespace {
+
+bool isComplexNan(const std::complex<double>& a) {
+  return std::isnan(a.real()) || std::isnan(a.imag());
```
Comment (jjsjann123) [inline csrc/ir/base_nodes.cpp:191]: naive question: do we care about matching nan on real/imag between the two value?

File: csrc/ir/base_nodes.cpp [hunk 4]
```diff
@@ -369,6 +427,23 @@ bool Expr::sameOp(const Expr* other) const {
...
 
+bool Expr::sameDefinition(const Expr* other) const {
+  if (!sameOp(other)) {
+    return false;
+  }
+  if (name() != other->name()) {
+    return false;
+  }
+  // Only check the input argument definitions to avoid infinite recursion
+  // because Val::sameDefinition will check the definition.
```
Comment (jjsjann123) [inline csrc/ir/base_nodes.cpp]: Does the comment mean we don't checking sameDefinition on outputs to avoid infinite recursion?!

The comment as-is reads a bit confusing, since we are calling Val::sameDefinition on input argument, while it says it's avoiding that.

File: csrc/tensor_view.cpp [hunk 5]
```diff
@@ -49,6 +49,21 @@ TensorView::TensorView(
 
 NVFUSER_DEFINE_CLONE(TensorView)
 
+bool TensorView::sameDefinition(const Val* other) const {
+  // Val::sameDefinition checks nullptr, dtype, vtype, and definition.
+  if (!Val::sameDefinition(other)) {
+    return false;
+  }
+  if (!other->isA<TensorView>()) {
```
Comment (jjsjann123) [inline csrc/tensor_view.cpp]: do we need to check this, since other has already passed Val::sameDefinition, which checks typeid already.

File: csrc/tensor_view.cpp [hunk 6]
```diff
@@ -49,6 +49,21 @@ TensorView::TensorView(
...
+    return false;
+  }
+  const TensorView* other_tv = other->as<TensorView>();
+  if (isCpuScalar() != other_tv->isCpuScalar()) {
+    return false;
+  }
+  return domain()->sameDefinition(other_tv->domain());
+}
```
Comment (jjsjann123) [inline csrc/tensor_view.cpp:61]: Out of curiosity, is scheduling supposed to be part of the check? I'm wondering what we do with sharding as well as other scheduling changes like compute at position.

Not that I'm advocating for that.

File: csrc/ir/nodes.cpp [hunk 7]
```diff
@@ -3517,6 +3520,46 @@ bool TensorDomain::hasGridBroadcast() const {
       });
 }
 
+bool TensorDomain::sameDefinition(const Val* other) const {
+  // Val::sameDefinition checks nullptr, dtype, vtype, and definition.
+  if (!Val::sameDefinition(other)) {
+    return false;
+  }
+  if (!other->isA<TensorDomain>()) {
```
Comment (jjsjann123) [inline csrc/ir/nodes.cpp]: nitpick, we've already checked sameDefinition, so this doesn't seem necessary.

File: csrc/ir/nodes.cpp [hunk 8]
```diff
@@ -3517,6 +3520,46 @@ bool TensorDomain::hasGridBroadcast() const {
...
+      return false;
+    }
+  }
+
+  // Check contiguity
+  if (contiguity_.size() != other_td->contiguity_.size()) {
+    return false;
+  }
+  return std::ranges::equal(contiguity_, other_td->contiguity_);
+}
```
Comment (jjsjann123) [inline csrc/ir/nodes.cpp]: From this, I imagine we are only handling non-scheduled / non-sharded fusion.

Wondering if we would want to check root domain as well.

File: csrc/ir/nodes.cpp [hunk 9]
```diff
@@ -3673,6 +3716,18 @@ std::vector<int64_t> TensorDomain::strideOrder() const {
...
 
+  // Operations like preprocessGroupedMatmulInputSf pad the logical domain to
+  // create the allocation domain. strideOrder only checks for permutations
+  // between logical and allocation domains.
+  bool is_complex_allocation = std::all_of(
+      allocation_domain_.begin(), allocation_domain_.end(), [](IterDomain* id) {
+        return id->extent()->definition() != nullptr;
+      });
+  NVF_CHECK(
+      !is_complex_allocation,
```
Comment (jjsjann123) [inline csrc/ir/nodes.cpp]: 👍

File: csrc/ir/nodes.cpp [hunk 10]
```diff
@@ -2716,45 +2716,48 @@ IterDomain::IterDomain(const IterDomain* src, IrCloner* ir_cloner)
...
+        getParallelType() == other_id->getParallelType() &&                 \
+        getIterType() == other_id->getIterType() &&                         \
+        hasPaddingToMultipleOfWarp() ==                                     \
+        other_id->hasPaddingToMultipleOfWarp() &&                           \
+        isClusteredBlockDim() == other_id->isClusteredBlockDim() &&         \
+        getMaybeSizeAfterPadding() == other_id->getMaybeSizeAfterPadding(); \
+  }
+
+ITERDOMAIN_SAME_FN(sameAs, Statement)
+ITERDOMAIN_SAME_FN(sameDefinition, Val)
```
Comment (jjsjann123) [inline csrc/ir/nodes.cpp]: dumb c++ question. What's the advantage of macro here, comparing to just define `Statement::sameAs` and `Val::sameDefinition` calling into `Statement::sameAs`.

**Review by jjsjann123 — COMMENTED** [review]
LGTM in general, some minor questions here and there.

File: csrc/ir/base_nodes.cpp (hunk 1 above)
Comment (rdspring1) [reply csrc/ir/base_nodes.cpp:143]: It is an assumption built into the `Val::sameAs` check today.

If a val doesn't have a definition, check its value if it exists. There isn't a val check if it has a definition.

If the assumption isn't true, then does the polymorphic value need to match the evaluated definition?

File: csrc/ir/base_nodes.cpp (hunk 3 above)
Comment (rdspring1) [reply csrc/ir/base_nodes.cpp:191]: We checked for this in `ScalarRecord`.

```cpp
if (value_.is<double>() && std::isnan(value_.as<double>()) &&
    std::isnan(child_ptr->value_.as<double>())) {
    return true;
}
```

https://github.com/NVIDIA/Fuser/blob/main/python/python_frontend/fusion_record.h#L2082-L2083

File: csrc/ir/base_nodes.cpp (hunk 4 above)
Comment (rdspring1) [reply csrc/ir/base_nodes.cpp]: Say Val A calls `sameDefinition`.  Val A calls `Expr::sameDefinition`. For this expr, Val A is the output. If you call check `sameDefinition` on the expr's outputs, it will call `sameDefinition` on Val A. Repeating forever.

File: csrc/ir/nodes.cpp (hunk 10 above)
Comment (rdspring1) [reply csrc/ir/nodes.cpp]: Minimizes code changes if `sameAs` or `sameDefinition` need to be modified.

Comment (rdspring1) [reply csrc/ir/nodes.cpp]: I don't know what this means:
```
define Statement::sameAs and Val::sameDefinition calling into Statement::sameAs
```

File: csrc/ir/base_nodes.cpp (hunk 2 above)
Comment (rdspring1) [reply csrc/ir/base_nodes.cpp:140]: This check means it might be the same val and requires further checking in `Val::sameAs` and `Val::sameDefinition`.

You're suggestion doesn't seem correct. If you return `false` in `notSameVal`, then, how will you know to check further in `Val::sameAs` or `Val::sameDefinition`.

Comment (rdspring1) [reply csrc/ir/base_nodes.cpp:140]: I'll rename it to `maybeSameVal`. It is a private helper function, so it cannot be mistakenly used.

File: csrc/tensor_view.cpp (hunk 6 above)
Comment (rdspring1) [reply csrc/tensor_view.cpp:61]: Scheduling is not checked.

File: csrc/ir/nodes.cpp (hunk 8 above)
Comment (rdspring1) [reply csrc/ir/nodes.cpp]: Can a non-scheduled fusion have a root domain?

Comment (rdspring1) [reply csrc/ir/nodes.cpp]: Adding more checks will slow down the LRU cache. It is already 5% more expensive because the nvfp4 operations create non-trivial allocation domains.

File: csrc/ir/base_nodes.cpp (hunk 4 above)
Comment (rdspring1) [reply csrc/ir/base_nodes.cpp]:
```cpp
  // Val::sameDefinition checks its definition expression. The value is the
  // output of its definition expression. Checking the output with
  // sameDefinition will create a cycle and cause infinite recursion, so only
  // check the input argument definitions. Fusion::sameDefinition starts from
  // each output and traverses up the entire Fusion DAG. 
```

**Comment by rdspring1** [comment]
!test

File: csrc/ir/base_nodes.cpp (hunk 2 above)
Comment (jjsjann123) [reply csrc/ir/base_nodes.cpp:140]: `maybeSameVal` is fine as well.

> You're suggestion doesn't seem correct. If you return false in notSameVal, then, how will you know to check further in Val::sameAs or Val::sameDefinition.

You are right... adding that maybe sounds better indeed.

File: csrc/ir/base_nodes.cpp (hunk 3 above)
Comment (jjsjann123) [reply csrc/ir/base_nodes.cpp:191]: sorry my question wasn't clear.

I meant for the user of isComplexNan, do they care about the two value being in the same form?

i.e. In `Val::sameDefinition`, if we have `this` as `1 + i(Nan)` and `other_val` being `Nan + i5`, they would be viewed as `sameDefinition` as well. But I'm not sure if that's problematic.

File: csrc/ir/nodes.cpp (hunk 8 above)
Comment (jjsjann123) [reply csrc/ir/nodes.cpp]: > Can a non-scheduled fusion have a root domain?

view op would give us root domain, since we use that to track the transformation.
For my own education, does scheduling add root domain? I thought it only manipulates loop.

File: csrc/ir/nodes.cpp (hunk 10 above)
Comment (jjsjann123) [reply csrc/ir/nodes.cpp]: > define Statement::sameAs and Val::sameDefinition calling into Statement::sameAs

i.e. just having:

```
bool IterDomain::sameAs(const Statement* other) const {
   // ... entire macro definition can go here
}

bool IterDomain::sameDefinition(const Val* other) const {
  return sameAs(other);
}
```

You don't have to change it. (I'm just not a big fan of macro)

File: csrc/ir/nodes.cpp (hunk 8 above)
Comment (rdspring1) [reply csrc/ir/nodes.cpp]: `rfactor`, a schedule operation, creates a root domain, right?

File: csrc/ir/base_nodes.cpp (hunk 3 above)
Comment (rdspring1) [reply csrc/ir/base_nodes.cpp:191]: `torch.allclose(x, y, equal_nan=True)` returns true if real or complex component is `nan` for either tensor.

```python
>>> x = torch.complex(torch.tensor([float('nan')]), torch.tensor([1.]))
>>> y = torch.complex(torch.tensor([10.]), torch.tensor([float('nan')]))
>>> x == y
tensor([False])
>>> torch.allclose(x, y)
False
>>> x
tensor([nan+1.j])
>>> y
tensor([10.+nanj])
>>> torch.allclose(x, y, equal_nan=True)
True
```

**Comment by rdspring1** [comment]
!test

File: csrc/ir/base_nodes.cpp (hunk 1 above)
Comment (jacobhinkle) [reply csrc/ir/base_nodes.cpp:143]: Gotcha. I think it makes sense that we don't check both value and definition here, so the rest of this comment is off-topic to this PR...

...but I wonder whether we should try to explicitly forbid setting both value and definition on a `Val`. I don't think we usually check for consistency in case there is both a `value()` and a `definition()` but we do such a validation when binding to derived values in ExpressionEvaluator:
https://github.com/NVIDIA/Fuser/blob/4a2cbf6bef7ef5677b6018ce9183176fa1994bc0/csrc/expr_evaluator.cpp#L194.

File: csrc/ir/base_nodes.cpp (hunk 3 above)
Comment (jjsjann123) [reply csrc/ir/base_nodes.cpp:191]: TIL. Thanks for that.

File: csrc/ir/nodes.cpp (hunk 8 above)
Comment (jjsjann123) [reply csrc/ir/nodes.cpp]: Keep forgetting about rfactor.
